packages/ui/src/components/Button/index.tsx

```diff
@@ -1,41 +1,104 @@
+import { I18nKey } from '@logto/phrases-ui';
 import classNames from 'classnames';
-import { ReactNode } from 'react';
+import { HTMLProps, ReactElement, ReactNode, useEffect, useRef, useState } from 'react';
+import { useTranslation } from 'react-i18next';
 
+import DangerousRaw from '../../../../console/src/components/DangerousRaw';
+import { Ring as Spinner } from '../../../../console/src/components/Spinner';
 import * as styles from './index.module.scss';
 
-export type Props = {
+export type ButtonType =
+  | 'primary'
+  | 'secondary'
+  | 'danger'
+  | 'outline'
+  | 'plain'
+  | 'default'
+  | 'branding';
+
+type BaseProps = Omit<HTMLProps<HTMLButtonElement>, 'type' | 'size' | 'title'> & {
   htmlType?: 'button' | 'submit' | 'reset';
+  type?: ButtonType;
+  size?: 'small' | 'medium' | 'large';
   isDisabled?: boolean;
-  className?: string;
-  children: ReactNode | Record<string, unknown>;
-  type?: 'primary' | 'secondary' | 'outline';
-  size?: 'small' | 'large';
-  onClick?: React.MouseEventHandler;
+  isLoading?: boolean;
+  loadingDelay?: number;
+};
+
+type TitleButtonProps = BaseProps & {
+  title: I18nKey | ReactElement<typeof DangerousRaw>;
+  icon?: ReactNode;
+  properties?: Record<string, string>;
+};
+
+type IconButtonProps = BaseProps & {
+  title?: I18nKey | ReactElement<typeof DangerousRaw>;
+  icon: ReactNode;
+  properties?: Record<string, string>;
 };
 
+export type Props = TitleButtonProps | IconButtonProps;
+
 const Button = ({
   htmlType = 'button',
-  type = 'primary',
-  size = 'large',
-  isDisabled,
+  type = 'default',
+  size = 'medium',
+  title,
+  properties,
+  icon,
   className,
-  children,
+  isDisabled = false,
+  isLoading = false,
+  loadingDelay = 500,
   onClick,
-}: Props) => (
-  <button
-    disabled={isDisabled}
-    className={classNames(
-      styles.button,
-      styles[type],
-      styles[size],
-      isDisabled && styles.disabled,
-      className
-    )}
-    type={htmlType}
-    onClick={onClick}
-  >
-    {children}
-  </button>
-);
+  ...rest
+}: Props) => {
+  const { t } = useTranslation();
+  const [showSpinner, setShowSpinner] = useState(false);
+  const timerRef = useRef<number>();
+
+  useEffect(() => {
+    // Delay showing the spinner after 'loadingDelay' milliseconds
+    if (isLoading) {
+      // DOC: To avoid typescript error
+      // eslint-disable-next-line @silverhand/fp/no-mutation
+      timerRef.current = window.setTimeout(() => {
+        setShowSpinner(true);
+      }, loadingDelay);
+    }
+
+    return () => {
+      clearTimeout(timerRef.current);
+      setShowSpinner(false);
+    };
+  }, [isLoading, loadingDelay]);
```

Comment:
Ditto:

Reply:
Hey, for this I see a lint issue. So I checked in StackOverflow, will this affect in any place? I tried explictly setting up type like this: `ReturnType<typeof setTimeout>` but it isn't worked.
Referece: https://stackoverflow.com/questions/55550096/ts2322-type-timeout-is-not-assignable-to-type-number-when-running-unit-te